### PR TITLE
UI: compare view: consolidated error handling, and details

### PR DESCRIPTION
**Pull request by jgehrcke**
Motivated by https://github.com/conbench/conbench/issues/1038.

This requires a few other PRs that are not yet merged.

Please let the feedback already come. Thanks!

Update: for review, posterity, paper trail I'd love to quickly share two screenshots from my local dev environment that show some of the user-facing aspects of the recent series of related patches. Note that in terms of visual/aesthetic aspects I tried to invest like 10 % of the effort to get to a -- well -- tolerable state quickly. I still think we are very much in the content-first stage, where it's super important to decide _which_ details to display roughly in which structure, and not yet so much about polishing visual detail. Yet, please keep any kind of feedback coming.

Screenshot from 2023-04-05 20-50-51

-----

Screenshot from 2023-04-05 20-41-38

When we look at these screenshots we see two other independent tasks that need to be taken care of:
- https://github.com/conbench/conbench/issues/1064
- https://github.com/conbench/conbench/issues/1063

### CODE REVIEW

File: conbench/app/compare.py [hunk 1]
```diff
@@ -142,78 +148,142 @@ def _get_plot(self, baseline, contender):
...
-        except ValueError:
-            baseline_id, contender_id = "unknown", "unknown"
-            comparisons, regressions, improvements = [], None, None
-            self.flash("Invalid contender and baseline.")
+    def get(self, compare_ids) -> str:
+        # Note that `compare_ids` is an unparsed string. Whether or not it
+        # contains known IDs, is for now only going to be known after having
+        # passed this to the API layer. Note that the two ideas that are
+        # encoded `compare_ids` can be either two run IDs, two batch IDs or two
+        # benchmark result IDs.
```
Comment (jgehrcke) [inline conbench/app/compare.py]: Typos... :-) Will polish.

File: conbench/app/compare.py [hunk 2]
```diff
@@ -142,78 +148,142 @@ def _get_plot(self, baseline, contender):
...
+            return self.error_page(
+                "No contender ID was provided. Expected format: <baseline-id>...<contender-id>"
+            )
+
+        (
+            comparison_results,
+            regression_count,
+            improvement_count,
+            error_string,
+        ) = self._compare(baseline_id=baseline_id, contender_id=contender_id)
```
Comment (jgehrcke) [inline conbench/app/compare.py:193]: I should add a comment that _compare() uses the API layer indirection.

**Comment by jgehrcke** [comment]
:ship: this.

I'd love to get your feedback @jonkeane @alistaire47 @austin3dickey! 

And I'd also love to try this out in staging, excited.

Not a huge fan of the visual appearance yet -- this is totally disputable -- see description of PR.

File: conbench/templates/compare-list.html [hunk 3]
```diff
@@ -17,31 +24,45 @@
...
+        </p>
       </div>
     </div>
-      {% endif %}
-    {% endif %}
+  </div>
+
+  <div class="col">
+    <div class="card text-bg-light mb-3">
+      <div class="card-header">Baseline / Reference</div>
```
Comment (austin3dickey) [inline conbench/templates/compare-list.html:53]: nit: Why not use just "Baseline" here?

File: conbench/templates/compare-list.html [hunk 4]
```diff
@@ -17,31 +24,45 @@
...
+{% if baseline_run and contender_run %}
+<div class="row row-cols-1 row-cols-md-2 g-4 mt-3 mb-5">
+  <div class="col">
+    <div class="card text-bg-dark mb-3" >
+      <div class="card-header">Contender</div>
+      <div class="card-body">
+        <h5 class="card-title"></h5>
+        <p class="card-text">
+          <a href="{{ url_for('app.run', run_id=contender_id) }}">Run {{ contender_id[:8] + "..."}}</a>
+          for commit {{ contender_run.commit.html_anchor_and_msg|safe }} submitted on  {{ contender_run.timestamp }}.
```
Comment (austin3dickey) [inline conbench/templates/compare-list.html:45]: small nit: I think "submitted at {timestamp}" reads a little smoother than "submitted on {timestamp}"

File: conbench/templates/compare-list.html [hunk 5]
```diff
@@ -17,31 +24,45 @@
...
-          <a href="{{ baseline_run.commit.url }}">{{ baseline_run.commit.display_message }}</a>
-          <br/>
-          Contender:
-          <a href="{{ contender_run.commit.url }}">{{ contender_run.commit.display_message }}</a>
-          </li>
-         </ul>
+{% if baseline_run and contender_run %}
+<div class="row row-cols-1 row-cols-md-2 g-4 mt-3 mb-5">
+  <div class="col">
+    <div class="card text-bg-dark mb-3" >
```
Comment (austin3dickey) [inline conbench/templates/compare-list.html:39]: nit: The dark card seems a little jarring, but it's not really a big deal. Maybe a lighter color?

File: conbench/app/compare.py [hunk 6]
```diff
@@ -142,78 +148,153 @@ def _get_plot(self, baseline, contender):
...
+
+        if error_string is not None:
+            return self.error_page(  # type: ignore
+                f"cannot perform comparison: {error_string}", alert_level="info"
+            )
+
+        if len(comparison_results) == 0:
+            return self.error_page(  # type: ignore
```
Comment (austin3dickey) [inline conbench/app/compare.py:201]: nice!

File: conbench/templates/compare-list.html (hunk 3 above)
Comment (jgehrcke) [reply conbench/templates/compare-list.html:53]: Very good question. We should use a single term, and that is Baseline (if that is what we use everywhere else). The `/ Reference` I ended up putting there quickly probably like a really pragmatic tooltip (having the itch to "explain" what the term "Baseline" means for someone who comes in freshly). This needs to change into a tooltip.

File: conbench/templates/compare-list.html (hunk 4 above)
Comment (jgehrcke) [reply conbench/templates/compare-list.html:45]: Oh. Thank you! :)

File: conbench/templates/compare-list.html (hunk 5 above)
Comment (jgehrcke) [reply conbench/templates/compare-list.html:39]: Yes. Not entirely happy with colors. I spent a tiny bit of time looking at 

- https://getbootstrap.com/docs/5.2/components/card/#card-styles
- https://getbootstrap.com/docs/5.2/components/card/#border
- https://getbootstrap.com/docs/5.2/helpers/color-background/

and wanted to find two contrasting representations, where the very coarse-grained visual impression already communicates that one thing is the "new guy who is going to be tested" (contender) and the other thing is the "established, critical thing" (the reference). 

I didn't find a quick way to nicely represent this visually.. red and green? no. Blue and green? no. Color at all? not really. Some other way? That's what I tried with

> Set a background-color with contrasting foreground color with [our .text-bg-{color} helpers](https://getbootstrap.com/docs/5.2/helpers/color-background/).

but yeah I also kind of gave up finding what I like most for now :)

File: conbench/templates/compare-list.html (hunk 4 above)
Comment (jgehrcke) [reply conbench/templates/compare-list.html:45]: still need to do that!

**Pull request merged**